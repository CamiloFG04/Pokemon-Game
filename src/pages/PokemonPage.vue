<template>
  <h1 v-if="!pokemon">Loading...</h1>
  <div v-else>
    <h1>Who is this Pok√©mon?</h1>
    <PokemonImage :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer"/>
    <template v-if="showAnswer">
      <h2 class="fade-in">{{message}}</h2>
      <button @click="newGame">Play again</button>
    </template>
  </div>

</template>

<script>
import PokemonImage from "@/components/PokemonImage.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import getPokemonOptions from "@/helpers/getPokemonOptions";
// console.log(getPokemonOptions());
export default {
  components: { PokemonImage, PokemonOptions },
  data() {
    return {
      pokemonArr:[],
      pokemon: null,
      showPokemon: false,
      showAnswer:false,
      message:null
    }
  },
  methods:{
    async mixPokemonArray(){
      this.pokemonArr = await getPokemonOptions()
      const rndInt = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArr[rndInt]
    },
    checkAnswer(pokemonId){
      this.showPokemon = true
      this.showAnswer = true
      this.message = pokemonId === this.pokemon.id ? `Correct, is ${this.pokemon.name.toUpperCase()}` : `Oops, was ${this.pokemon.name.toUpperCase()}`
    },
    newGame(){
      this.pokemon = null
      this.pokemonArr = []
      this.mixPokemonArray()
      this.showPokemon = false
      this.showAnswer = false
      this.message = null
    }
  },
  mounted(){
    this.mixPokemonArray()
  }

}
</script>
