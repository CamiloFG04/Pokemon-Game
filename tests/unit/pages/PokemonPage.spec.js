import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { pokemons } from "../mocks/pokemons.mock";

describe('PokemonPage Component', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    });

    test('must call mixPokemonArray when mounting', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods,'mixPokemonArray')
        shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled();
    });

    test('must be match with snapshot when pokemons are loaded', () => {
        const wrapper = shallowMount(PokemonPage,{
            data(){
                return{
                    pokemonArr:pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer:false,
                    message:''
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('must be show PokemonImage and PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage,{
            data(){
                return{
                    pokemonArr:pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer:false,
                    message:''
                }
            }
        })
        const image = wrapper.getComponent({name:'PokemonImage'});
        const options = wrapper.getComponent({name:'PokemonOptions'});
        expect(image).toBeTruthy()
        expect(options).toBeTruthy()
        expect(image.attributes('pokemonid')).toBe("1")
        expect(options.attributes('pokemons')).toBeTruthy()
    });

    test('tests with checkAnswer', async() => {
        const wrapper = shallowMount(PokemonPage,{
            data(){
                return{
                    pokemonArr:pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer:false,
                    message:''
                }
            }
        })
        await wrapper.vm.checkAnswer(1)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correct, is ${pokemons[0].name.toUpperCase()}`)
        expect(wrapper.vm.showPokemon).toBeTruthy()

        await wrapper.vm.checkAnswer(10)

        expect(wrapper.vm.message).toBe(`Oops, was ${pokemons[0].name.toUpperCase()}`)


    });
});