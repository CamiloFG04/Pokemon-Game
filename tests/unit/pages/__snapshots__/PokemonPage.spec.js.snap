// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component must be match with snapshot when pokemons are loaded 1`] = `
<div>
  <h1>Who is this Pok√©mon?</h1>
  <pokemon-image-stub pokemonid="1" showpokemon="false"></pokemon-image-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
